docs/rfc/rfc-026-p2p-bad-peers-checktx.md

```diff
@@ -0,0 +1,375 @@
+we might need more information on the actual reason for rejection. This could be done by an additional set of response codes provided by the application. 
+
+### 2. Choosing the peer to ban
+
+Each transaction gossiped contains the ID of the peer that sent that transaction. Upon receiving a transaction,
+a node saves the peer ID of the peer(s) that have sent it. As each peer had to have
+run `CheckTx` on this transaction before adding it to its own mempool, we can assume this peer
+can be held accountable for the validity of transactions it gossips. Invalid transactions are kept only in the mempool
+cache and thus not gossiped. 
+
+*Transactions received by users*
+
+For transactions submitted via `boradcastTxCommit`, the `SenderID` field is empty. 
+
+**Note**  Do we have mechanisms in place to handle cases when `broadcastTxCommit` submits
+failing transactions (can this be a form of attack)?
+
+### 3. Attack scenarios
+
+While an attack by simply banning peers on faling `CheckTx` is hard to imagine, as the incentive for doing so is not clear, there are considerations with regards to the current mempool gossip implementation.
+
+Should we keep transactions that could never have been valid in the cache? Assuming that receiving such transactions is rare, and the peer that sent them is banned, do we need to occupy space in the mempool cache with these transactions?
+
+
+## Implementation considerations
+
+When a transaction failes `CheckTx`,it is not stored in the mempool but **can** be stored in the cache. If it is in the cache, it cannot be resubmitted again (as it will be discovered in the cache and not checked again). These two scenarios require a different implementation of banning in case `CheckTx` failed. 
+
+In both cases we need to keep track of the peers that sent invalid transactions. If invalid transactions are  cached, we also need to keep track of the `CheckTx` response code for each transaction. Currently the `ResponseCheckTx` code is checked in `resCbFirstTime` of the mempool. If  invalid transactions are kept in the cache, the check is ran only when a transaction is 
+seen for the first time. Afterwards, the  transaction is cached, to avoid running `CheckTx` on transactions already checked. 
+Thus when a transaction is received from a peer, if it is in the cache,
+`CheckTx` is not ran again, but the peers' ID is addded to the list of peers who sent this particular transaction.
+These transactions are rechecked once a block is committed to verify that they are still valid. 
+
+If invalid transactions are not kept in the cache, they can be resubmitted multiple times, and `CheckTx` will be executed on them upon submission therefore we do not need to remember the previous response codes for these transactions.
+
+Currently there are two ways to implement peer banning based on the result of `CheckTx`:
+
+1. Introduce banning when transactions start to be processed
+2. Adapt the recheck logic to support this
```

Comment:
What would be an example of a `reCheckTx` returning a code >1. If the TX is in the local mempool, the same node ran `CheckTx` earlier and passed the check.
It looks pretty much like a bug in the application's code to me.